File: hosting-service/src/lib/redirects.ts
import { readFile } from 'fs/promises';
import { existsSync } from 'fs';

export interface RedirectRule {
  from: string;
  to: string;
  status: number;
  force: boolean;
  conditions?: {
    country?: string[];
    language?: string[];
    role?: string[];
    cookie?: string[];
  };
  // For pattern matching
  fromPattern?: RegExp;
  fromParams?: string[]; // Named parameters from the pattern
  queryParams?: Record<string, string>; // Expected query parameters
}

export interface RedirectMatch {
  rule: RedirectRule;
  targetPath: string;
  status: number;
}

// Maximum number of redirect rules to prevent DoS attacks
const MAX_REDIRECT_RULES = 1000;

/**
 * Parse a _redirects file into an array of redirect rules
 */
export function parseRedirectsFile(content: string): RedirectRule[] {
  const lines = content.split('\n');
  const rules: RedirectRule[] = [];

  for (let lineNum = 0; lineNum < lines.length; lineNum++) {
    const lineRaw = lines[lineNum];
    if (!lineRaw) continue;

    const line = lineRaw.trim();

    // Skip empty lines and comments
    if (!line || line.startsWith('#')) {
      continue;
    }

    // Enforce max rules limit
    if (rules.length >= MAX_REDIRECT_RULES) {
      console.warn(`Redirect rules limit reached (${MAX_REDIRECT_RULES}), ignoring remaining rules`);
      break;
    }

    try {
      const rule = parseRedirectLine(line);
      if (rule && rule.fromPattern) {
        rules.push(rule);
      }
    } catch (err) {
      console.warn(`Failed to parse redirect rule on line ${lineNum + 1}: ${line}`, err);
    }
  }

  return rules;
}

/**
 * Parse a single redirect rule line
 * Format: /from [query_params] /to [status] [conditions]
 */
function parseRedirectLine(line: string): RedirectRule | null {
  // Split by whitespace, but respect quoted strings (though not commonly used)
  const parts = line.split(/\s+/);
  
  if (parts.length < 2) {
    return null;
  }

  let idx = 0;
  const from = parts[idx++];
  
  if (!from) {
    return null;
  }
  
  let status = 301; // Default status
  let force = false;
  const conditions: NonNullable<RedirectRule['conditions']> = {};
  const queryParams: Record<string, string> = {};
  
  // Parse query parameters that come before the destination path
  // They look like: key=:value (and don't start with /)
  while (idx < parts.length) {
    const part = parts[idx];
    if (!part) {
      idx++;
      continue;
    }
    
    // If it starts with / or http, it's the destination path
    if (part.startsWith('/') || part.startsWith('http://') || part.startsWith('https://')) {
      break;
    }
    
    // If it contains = and comes before the destination, it's a query param
    if (part.includes('=')) {
      const splitIndex = part.indexOf('=');
      const key = part.slice(0, splitIndex);
      const value = part.slice(splitIndex + 1);
      
      if (key && value) {
        queryParams[key] = value;
      }
      idx++;
    } else {
      // Not a query param, must be destination or something else
      break;
    }
  }
  
  // Next part should be the destination
  if (idx >= parts.length) {
    return null;
  }
  
  const to = parts[idx++];
  if (!to) {
    return null;
  }

  // Parse remaining parts for status code and conditions
  for (let i = idx; i < parts.length; i++) {
    const part = parts[i];
    
    if (!part) continue;
    
    // Check for status code (with optional ! for force)
    if (/^\d+!?$/.test(part)) {
      if (part.endsWith('!')) {
        force = true;
        status = parseInt(part.slice(0, -1));
      } else {
        status = parseInt(part);
      }
      continue;
    }

    // Check for condition parameters (Country=, Language=, Role=, Cookie=)
    if (part.includes('=')) {
      const splitIndex = part.indexOf('=');
      const key = part.slice(0, splitIndex);
      const value = part.slice(splitIndex + 1);
      
      if (!key || !value) continue;
      
      const keyLower = key.toLowerCase();
      
      if (keyLower === 'country') {
        conditions.country = value.split(',').map(v => v.trim().toLowerCase());
      } else if (keyLower === 'language') {
        conditions.language = value.split(',').map(v => v.trim().toLowerCase());
      } else if (keyLower === 'role') {
        conditions.role = value.split(',').map(v => v.trim());
      } else if (keyLower === 'cookie') {
        conditions.cookie = value.split(',').map(v => v.trim().toLowerCase());
      }
    }
  }

  // Parse the 'from' pattern
  const { pattern, params } = convertPathToRegex(from);

  return {
    from,
    to,
    status,
    force,
    conditions: Object.keys(conditions).length > 0 ? conditions : undefined,
    queryParams: Object.keys(queryParams).length > 0 ? queryParams : undefined,
    fromPattern: pattern,
    fromParams: params,
  };
}

/**
 * Convert a path pattern with placeholders and splats to a regex
 * Examples:
 *   /blog/:year/:month/:day -> captures year, month, day
 *   /news/* -> captures splat
 */
function convertPathToRegex(pattern: string): { pattern: RegExp; params: string[] } {
  const params: string[] = [];
  let regexStr = '^';
  
  // Split by query string if present
  const pathPart = pattern.split('?')[0] || pattern;
  
  // Escape special regex characters except * and :
  let escaped = pathPart.replace(/[.+^${}()|[\]\\]/g, '\\$&');
  
  // Replace :param with named capture groups
  escaped = escaped.replace(/:([a-zA-Z_][a-zA-Z0-9_]*)/g, (match, paramName) => {
    params.push(paramName);
    // Match path segment (everything except / and ?)
    return '([^/?]+)';
  });
  
  // Replace * with splat capture (matches everything including /)
  if (escaped.includes('*')) {
    escaped = escaped.replace(/\*/g, '(.*)');
    params.push('splat');
  }
  
  regexStr += escaped;
  
  // Make trailing slash optional
  if (!regexStr.endsWith('.*')) {
    regexStr += '/?';
  }
  
  regexStr += '$';
  
  return {
    pattern: new RegExp(regexStr),
    params,
  };
}

/**
 * Match a request path against redirect rules with loop detection
 */
export function matchRedirectRule(
  requestPath: string,
  rules: RedirectRule[],
  context?: {
    queryParams?: Record<string, string>;
    headers?: Record<string, string>;
    cookies?: Record<string, string>;
  },
  visitedPaths: Set<string> = new Set()
): RedirectMatch | null {
  // Normalize path: ensure leading slash, remove trailing slash (except for root)
  let normalizedPath = requestPath.startsWith('/') ? requestPath : `/${requestPath}`;

  // Detect redirect loops
  if (visitedPaths.has(normalizedPath)) {
    console.warn(`Redirect loop detected for path: ${normalizedPath}`);
    return null;
  }

  // Track this path to detect loops
  visitedPaths.add(normalizedPath);

  // Limit redirect chain depth to 10
  if (visitedPaths.size > 10) {
    console.warn(`Redirect chain too deep (>10) for path: ${normalizedPath}`);
    return null;
  }

  for (const rule of rules) {
    // Check query parameter conditions first (if any)
    if (rule.queryParams) {
      // If rule requires query params but none provided, skip this rule
      if (!context?.queryParams) {
        continue;
      }

      // Check that all required query params are present
      // The value in rule.queryParams is either a literal or a placeholder (:name)
      const queryMatches = Object.entries(rule.queryParams).every(([key, expectedValue]) => {
        const actualValue = context.queryParams?.[key];

        // Query param must exist
        if (actualValue === undefined) {
          return false;
        }

        // If expected value is a placeholder (:name), any value is acceptable
        // If it's a literal, it must match exactly
        if (expectedValue && !expectedValue.startsWith(':')) {
          return actualValue === expectedValue;
        }

        return true;
      });

      if (!queryMatches) {
        continue;
      }
    }

    // Check conditional redirects (country, language, role, cookie)
    if (rule.conditions) {
      if (rule.conditions.country && context?.headers) {
        const cfCountry = context.headers['cf-ipcountry'];
        const xCountry = context.headers['x-country'];
        const country = (cfCountry?.toLowerCase() || xCountry?.toLowerCase());
        if (!country || !rule.conditions.country.includes(country)) {
          continue;
        }
      }

      if (rule.conditions.language && context?.headers) {
        const acceptLang = context.headers['accept-language'];
        if (!acceptLang) {
          continue;
        }
        // Parse accept-language header (simplified)
        const langs = acceptLang.split(',').map(l => {
          const langPart = l.split(';')[0];
          return langPart ? langPart.trim().toLowerCase() : '';
        }).filter(l => l !== '');
        const hasMatch = rule.conditions.language.some(lang => 
          langs.some(l => l === lang || l.startsWith(lang + '-'))
        );
        if (!hasMatch) {
          continue;
        }
      }

      if (rule.conditions.cookie && context?.cookies) {
        const hasCookie = rule.conditions.cookie.some(cookieName => 
          context.cookies && cookieName in context.cookies
        );
        if (!hasCookie) {
          continue;
        }
      }

      // Role-based redirects would need JWT verification - skip for now
      if (rule.conditions.role) {
        continue;
      }
    }

    // Match the path pattern
    const match = rule.fromPattern?.exec(normalizedPath);
    if (!match) {
      continue;
    }

    // Build the target path by replacing placeholders
    let targetPath = rule.to;

    // Replace captured parameters (with URL encoding)
    if (rule.fromParams && match.length > 1) {
      for (let i = 0; i < rule.fromParams.length; i++) {
        const paramName = rule.fromParams[i];
        const paramValue = match[i + 1];

        if (!paramName || !paramValue) continue;

        // URL encode captured values to prevent invalid URLs
        const encodedValue = encodeURIComponent(paramValue);

        if (paramName === 'splat') {
          // For splats, preserve slashes by re-decoding them
          const splatValue = encodedValue.replace(/%2F/g, '/');
          targetPath = targetPath.replace(':splat', splatValue);
        } else {
          targetPath = targetPath.replace(`:${paramName}`, encodedValue);
        }
      }
    }

    // Handle query parameter replacements (with URL encoding)
    if (rule.queryParams && context?.queryParams) {
      for (const [key, placeholder] of Object.entries(rule.queryParams)) {
        const actualValue = context.queryParams[key];
        if (actualValue && placeholder && placeholder.startsWith(':')) {
          const paramName = placeholder.slice(1);
          if (paramName) {
            // URL encode query parameter values
            const encodedValue = encodeURIComponent(actualValue);
            targetPath = targetPath.replace(`:${paramName}`, encodedValue);
          }
        }
      }
    }

    // Preserve query string for 200, 301, 302 redirects (unless target already has one)
    if ([200, 301, 302].includes(rule.status) && context?.queryParams && !targetPath.includes('?')) {
      const queryString = Object.entries(context.queryParams)
        .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
        .join('&');
      if (queryString) {
        targetPath += `?${queryString}`;
      }
    }

    return {
      rule,
      targetPath,
      status: rule.status,
    };
  }

  return null;
}

/**
 * Load redirect rules from a cached site
 */
export async function loadRedirectRules(did: string, rkey: string): Promise<RedirectRule[]> {
  const CACHE_DIR = process.env.CACHE_DIR || './cache/sites';
  const redirectsPath = `${CACHE_DIR}/${did}/${rkey}/_redirects`;
  
  if (!existsSync(redirectsPath)) {
    return [];
  }

  try {
    const content = await readFile(redirectsPath, 'utf-8');
    return parseRedirectsFile(content);
  } catch (err) {
    console.error('Failed to load _redirects file', err);
    return [];
  }
}

/**
 * Parse cookies from Cookie header
 */
export function parseCookies(cookieHeader?: string): Record<string, string> {
  if (!cookieHeader) return {};
  
  const cookies: Record<string, string> = {};
  const parts = cookieHeader.split(';');
  
  for (const part of parts) {
    const [key, ...valueParts] = part.split('=');
    if (key && valueParts.length > 0) {
      cookies[key.trim()] = valueParts.join('=').trim();
    }
  }
  
  return cookies;
}

/**
 * Parse query string into object
 */
export function parseQueryString(url: string): Record<string, string> {
  const queryStart = url.indexOf('?');
  if (queryStart === -1) return {};
  
  const queryString = url.slice(queryStart + 1);
  const params: Record<string, string> = {};
  
  for (const pair of queryString.split('&')) {
    const [key, value] = pair.split('=');
    if (key) {
      params[decodeURIComponent(key)] = value ? decodeURIComponent(value) : '';
    }
  }
  
  return params;
}

File: rust-cli/src/redirects.rs
use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

/// Maximum number of redirect rules to prevent DoS attacks
const MAX_REDIRECT_RULES: usize = 1000;

#[derive(Debug, Clone)]
pub struct RedirectRule {
    #[allow(dead_code)]
    pub from: String,
    pub to: String,
    pub status: u16,
    #[allow(dead_code)]
    pub force: bool,
    pub from_pattern: Regex,
    pub from_params: Vec<String>,
    pub query_params: Option<HashMap<String, String>>,
}

#[derive(Debug)]
pub struct RedirectMatch {
    pub target_path: String,
    pub status: u16,
    pub force: bool,
}

/// Parse a _redirects file into an array of redirect rules
pub fn parse_redirects_file(content: &str) -> Vec<RedirectRule> {
    let lines = content.lines();
    let mut rules = Vec::new();

    for (line_num, line_raw) in lines.enumerate() {
        if line_raw.trim().is_empty() || line_raw.trim().starts_with('#') {
            continue;
        }

        // Enforce max rules limit
        if rules.len() >= MAX_REDIRECT_RULES {
            eprintln!(
                "Redirect rules limit reached ({}), ignoring remaining rules",
                MAX_REDIRECT_RULES
            );
            break;
        }

        match parse_redirect_line(line_raw.trim()) {
            Ok(Some(rule)) => rules.push(rule),
            Ok(None) => continue,
            Err(e) => {
                eprintln!(
                    "Failed to parse redirect rule on line {}: {} ({})",
                    line_num + 1,
                    line_raw,
                    e
                );
            }
        }
    }

    rules
}

/// Parse a single redirect rule line
/// Format: /from [query_params] /to [status] [conditions]
fn parse_redirect_line(line: &str) -> Result<Option<RedirectRule>, String> {
    let parts: Vec<&str> = line.split_whitespace().collect();

    if parts.len() < 2 {
        return Ok(None);
    }

    let mut idx = 0;
    let from = parts[idx];
    idx += 1;

    let mut status = 301; // Default status
    let mut force = false;
    let mut query_params: HashMap<String, String> = HashMap::new();

    // Parse query parameters that come before the destination path
    while idx < parts.len() {
        let part = parts[idx];

        // If it starts with / or http, it's the destination path
        if part.starts_with('/') || part.starts_with("http://") || part.starts_with("https://") {
            break;
        }

        // If it contains = and comes before the destination, it's a query param
        if part.contains('=') {
            let split_index = part.find('=').unwrap();
            let key = &part[..split_index];
            let value = &part[split_index + 1..];

            if !key.is_empty() && !value.is_empty() {
                query_params.insert(key.to_string(), value.to_string());
            }
            idx += 1;
        } else {
            break;
        }
    }

    // Next part should be the destination
    if idx >= parts.len() {
        return Ok(None);
    }

    let to = parts[idx];
    idx += 1;

    // Parse remaining parts for status code
    for part in parts.iter().skip(idx) {
        // Check for status code (with optional ! for force)
        if let Some(stripped) = part.strip_suffix('!') {
            if let Ok(s) = stripped.parse::<u16>() {
                force = true;
                status = s;
            }
        } else if let Ok(s) = part.parse::<u16>() {
            status = s;
        }
        // Note: We're ignoring conditional redirects (Country, Language, Cookie, Role) for now
        // They can be added later if needed
    }

    // Parse the 'from' pattern
    let (pattern, params) = convert_path_to_regex(from)?;

    Ok(Some(RedirectRule {
        from: from.to_string(),
        to: to.to_string(),
        status,
        force,
        from_pattern: pattern,
        from_params: params,
        query_params: if query_params.is_empty() {
            None
        } else {
            Some(query_params)
        },
    }))
}

/// Convert a path pattern with placeholders and splats to a regex
/// Examples:
///   /blog/:year/:month/:day -> captures year, month, day
///   /news/* -> captures splat
fn convert_path_to_regex(pattern: &str) -> Result<(Regex, Vec<String>), String> {
    let mut params = Vec::new();
    let mut regex_str = String::from("^");

    // Split by query string if present
    let path_part = pattern.split('?').next().unwrap_or(pattern);

    // Escape special regex characters except * and :
    let mut escaped = String::new();
    for ch in path_part.chars() {
        match ch {
            '.' | '+' | '^' | '$' | '{' | '}' | '(' | ')' | '|' | '[' | ']' | '\\' => {
                escaped.push('\\');
                escaped.push(ch);
            }
            _ => escaped.push(ch),
        }
    }

    // Replace :param with named capture groups
    let param_regex = Regex::new(r":([a-zA-Z_][a-zA-Z0-9_]*)").map_err(|e| e.to_string())?;
    let mut last_end = 0;
    let mut result = String::new();

    for cap in param_regex.captures_iter(&escaped) {
        let m = cap.get(0).unwrap();
        result.push_str(&escaped[last_end..m.start()]);
        result.push_str("([^/?]+)");
        params.push(cap[1].to_string());
        last_end = m.end();
    }
    result.push_str(&escaped[last_end..]);
    escaped = result;

    // Replace * with splat capture
    if escaped.contains('*') {
        escaped = escaped.replace('*', "(.*)");
        params.push("splat".to_string());
    }

    regex_str.push_str(&escaped);

    // Make trailing slash optional
    if !regex_str.ends_with(".*") {
        regex_str.push_str("/?");
    }

    regex_str.push('$');

    let pattern = Regex::new(&regex_str).map_err(|e| e.to_string())?;

    Ok((pattern, params))
}

/// Match a request path against redirect rules
pub fn match_redirect_rule(
    request_path: &str,
    rules: &[RedirectRule],
    query_params: Option<&HashMap<String, String>>,
) -> Option<RedirectMatch> {
    // Normalize path: ensure leading slash
    let normalized_path = if request_path.starts_with('/') {
        request_path.to_string()
    } else {
        format!("/{}", request_path)
    };

    for rule in rules {
        // Check query parameter conditions first (if any)
        if let Some(required_params) = &rule.query_params {
            if let Some(actual_params) = query_params {
                let query_matches = required_params.iter().all(|(key, expected_value)| {
                    if let Some(actual_value) = actual_params.get(key) {
                        // If expected value is a placeholder (:name), any value is acceptable
                        if expected_value.starts_with(':') {
                            return true;
                        }
                        // Otherwise it must match exactly
                        actual_value == expected_value
                    } else {
                        false
                    }
                });

                if !query_matches {
                    continue;
                }
            } else {
                // Rule requires query params but none provided
                continue;
            }
        }

        // Match the path pattern
        if let Some(captures) = rule.from_pattern.captures(&normalized_path) {
            let mut target_path = rule.to.clone();

            // Replace captured parameters
            for (i, param_name) in rule.from_params.iter().enumerate() {
                if let Some(param_value) = captures.get(i + 1) {
                    let value = param_value.as_str();

                    if param_name == "splat" {
                        target_path = target_path.replace(":splat", value);
                    } else {
                        target_path = target_path.replace(&format!(":{}", param_name), value);
                    }
                }
            }

            // Handle query parameter replacements
            if let Some(required_params) = &rule.query_params {
                if let Some(actual_params) = query_params {
                    for (key, placeholder) in required_params {
                        if placeholder.starts_with(':') {
                            if let Some(actual_value) = actual_params.get(key) {
                                let param_name = &placeholder[1..];
                                target_path = target_path.replace(
                                    &format!(":{}", param_name),
                                    actual_value,
                                );
                            }
                        }
                    }
                }
            }

            // Preserve query string for 200, 301, 302 redirects (unless target already has one)
            if [200, 301, 302].contains(&rule.status)
                && query_params.is_some()
                && !target_path.contains('?')
            {
                if let Some(params) = query_params {
                    if !params.is_empty() {
                        let query_string: String = params
                            .iter()
                            .map(|(k, v)| format!("{}={}", k, v))
                            .collect::<Vec<_>>()
                            .join("&");
                        target_path = format!("{}?{}", target_path, query_string);
                    }
                }
            }

            return Some(RedirectMatch {
                target_path,
                status: rule.status,
                force: rule.force,
            });
        }
    }

    None
}

/// Load redirect rules from a _redirects file
pub fn load_redirect_rules(directory: &Path) -> Vec<RedirectRule> {
    let redirects_path = directory.join("_redirects");

    if !redirects_path.exists() {
        return Vec::new();
    }

    match fs::read_to_string(&redirects_path) {
        Ok(content) => parse_redirects_file(&content),
        Err(e) => {
            eprintln!("Failed to load _redirects file: {}", e);
            Vec::new()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_simple_redirect() {
        let content = "/old-path /new-path";
        let rules = parse_redirects_file(content);
        assert_eq!(rules.len(), 1);
        assert_eq!(rules[0].from, "/old-path");
        assert_eq!(rules[0].to, "/new-path");
        assert_eq!(rules[0].status, 301);
        assert!(!rules[0].force);
    }

    #[test]
    fn test_parse_with_status() {
        let content = "/temp /target 302";
        let rules = parse_redirects_file(content);
        assert_eq!(rules[0].status, 302);
    }

    #[test]
    fn test_parse_force_redirect() {
        let content = "/force /target 301!";
        let rules = parse_redirects_file(content);
        assert!(rules[0].force);
    }

    #[test]
    fn test_match_exact_path() {
        let rules = parse_redirects_file("/old-path /new-path");
        let m = match_redirect_rule("/old-path", &rules, None);
        assert!(m.is_some());
        assert_eq!(m.unwrap().target_path, "/new-path");
    }

    #[test]
    fn test_match_splat() {
        let rules = parse_redirects_file("/news/* /blog/:splat");
        let m = match_redirect_rule("/news/2024/01/15/post", &rules, None);
        assert!(m.is_some());
        assert_eq!(m.unwrap().target_path, "/blog/2024/01/15/post");
    }

    #[test]
    fn test_match_placeholders() {
        let rules = parse_redirects_file("/blog/:year/:month/:day /posts/:year-:month-:day");
        let m = match_redirect_rule("/blog/2024/01/15", &rules, None);
        assert!(m.is_some());
        assert_eq!(m.unwrap().target_path, "/posts/2024-01-15");
    }
}
